shell/platform/darwin/macos/framework/Source/FLEEngine.mm

```diff
@@ -332,60 +332,54 @@ - (void)shutDownEngine {
 #pragma mark - FlutterBinaryMessenger
 
 - (void)sendOnChannel:(nonnull NSString*)channel message:(nullable NSData*)message {
-  FlutterPlatformMessage platformMessage = {
-      .struct_size = sizeof(FlutterPlatformMessage),
-      .channel = [channel UTF8String],
-      .message = static_cast<const uint8_t*>(message.bytes),
-      .message_size = message.length,
-  };
-
-  FlutterEngineResult result = FlutterEngineSendPlatformMessage(_engine, &platformMessage);
-  if (result != kSuccess) {
-    NSLog(@"Failed to send message to Flutter engine on channel '%@' (%d).", channel, result);
-  }
+  [self sendOnChannel:channel message:message binaryReply:nil];
 }
 
 - (void)sendOnChannel:(NSString*)channel
               message:(NSData* _Nullable)message
           binaryReply:(FlutterBinaryReply _Nullable)callback {
-  struct Captures {
-    FlutterBinaryReply reply;
-  };
-  auto captures = std::make_unique<Captures>();
-  captures->reply = callback;
-  auto message_reply = [](const uint8_t* data, size_t data_size, void* user_data) {
-    auto captures = reinterpret_cast<Captures*>(user_data);
-    NSData* reply_data = [NSData dataWithBytes:(void*)data length:data_size];
-    captures->reply(reply_data);
-    delete captures;
-  };
-
-  FlutterPlatformMessageResponseHandle* response_handle = nullptr;
-  FlutterEngineResult result = FlutterPlatformMessageCreateResponseHandle(
-      _engine, message_reply, captures.get(), &response_handle);
-  if (result != kSuccess) {
-    NSLog(@"Failed to create a FlutterPlatformMessageResponseHandle");
-    return;
-  }
-  captures.release();
-
   FlutterPlatformMessage platformMessage = {
       .struct_size = sizeof(FlutterPlatformMessage),
       .channel = [channel UTF8String],
       .message = static_cast<const uint8_t*>(message.bytes),
       .message_size = message.length,
-      .response_handle = response_handle,
   };
+  FlutterPlatformMessageResponseHandle* response_handle = nullptr;
+
+  if (callback) {
+    struct Captures {
+      FlutterBinaryReply reply;
+    };
+    auto captures = std::make_unique<Captures>();
+    captures->reply = callback;
+    auto message_reply = [](const uint8_t* data, size_t data_size, void* user_data) {
+      auto captures = reinterpret_cast<Captures*>(user_data);
+      NSData* reply_data = [NSData dataWithBytes:(void*)data length:data_size];
+      captures->reply(reply_data);
+      delete captures;
+    };
+
+    FlutterEngineResult result = FlutterPlatformMessageCreateResponseHandle(
+        _engine, message_reply, captures.get(), &response_handle);
```

Comment:
Same concern here as in the other patch, now that I look at this again: what is guaranteeing the lifetime of `message_reply`? Does the spec guarantee this for non-capturing lambas?

It might be clearer, and easier to ensure there's no lifetime problem, if you use a static function instead of a lambda, but a comment with a spec reference would work too (assuming it is indeed correct as written.)

Reply:
`message_reply` is just a pointer to an anonymous function and has no lifecycle issues you described. [See section 8.1.5.1 Section 6 of the C++ Standard](http://www.open-std.org/JTC1/SC22/WG21/docs/papers/2017/n4659.pdf). In fact, if any captures were specified (by mistake), the implicit cast when the C routine is called would fail and you would get a compiler error.

Reply:
Thanks. I thought that I remembered that non-capturing lambdas were handled differently, but was having troubling finding where the confirmation was.